### PR TITLE
Other changes needed to support both Rack 2 and 3

### PR DESCRIPTION
**Pull request by schinery**

### CODE REVIEW

File: lib/grape/http/headers.rb [hunk 1]
```diff
@@ -62,6 +62,13 @@ module Headers
       def self.find_supported_method(route_method)
         Grape::Http::Headers::SUPPORTED_METHODS.detect { |supported_method| supported_method.casecmp(route_method).zero? }
       end
+
+      def self.rackified_header(key)
+        return key if Gem::Version.new(Rack.release) < Gem::Version.new('3')
+
+        # .dup because the key is frozen
+        key.dup.downcase!
```
Comment (dblock) [inline lib/grape/http/headers.rb]: Just return `key.downcase` instead of downcasing it in place.

File: lib/grape/http/headers.rb [hunk 2]
```diff
@@ -62,6 +62,13 @@ module Headers
       def self.find_supported_method(route_method)
         Grape::Http::Headers::SUPPORTED_METHODS.detect { |supported_method| supported_method.casecmp(route_method).zero? }
       end
+
+      def self.rackified_header(key)
```
Comment (dblock) [inline lib/grape/http/headers.rb]: Name it `header_name`? But I think this method shouldn't exist. It's very much an internal thing of what the header is stored as. We have a `transform` method that should `downcase` it.

File: lib/grape/request.rb [hunk 3]
```diff
@@ -47,7 +47,8 @@ def build_headers
     end
 
     def transform_header(header)
-      -header[5..].split('_').each(&:capitalize!).join('-')
+      transformed = -header[5..].split('_').each(&:capitalize!).join('-')
+      Grape::Http::Headers.rackified_header(transformed)
```
Comment (dblock) [inline lib/grape/request.rb]: Just downcase it here.

File: spec/grape/api/custom_validations_spec.rb [hunk 4]
```diff
@@ -162,9 +162,15 @@ def validate(request)
...
+
           # check if user is admin or not
           # as an example get a token from request and check if it's admin or not
-          raise Grape::Exceptions::Validation.new(params: @attrs, message: 'Can not set Admin only field.') unless request.headers['X-Access-Token'] == 'admin'
+          header = if Gem::Version.new(Rack.release) < Gem::Version.new('3')
+                     'X-Access-Token'
+                   else
+                     'x-access-token'
+                   end
+          raise Grape::Exceptions::Validation.new(params: @attrs, message: 'Can not set Admin only field.') unless request.headers[header] == 'admin'
```
Comment (dblock) [inline spec/grape/api/custom_validations_spec.rb:173]: Would it make more sense if `request.headers` became a dict with case-insensitive keys instead of downcasing everything?

**Comment by dblock** [comment]
tl;dr would this work?

- Grape would downcase headers it returns with Rack 3 before sending
- `headers` would become a case-insensitive collection in Rack 3

**Comment by schinery** [comment]
> tl;dr would this work?
> 
> * Grape would downcase headers it returns with Rack 3 before sending
> * `headers` would become a case-insensitive collection in Rack 3

@dblock do you have an example of what you are thinking for both of these?

File: lib/grape/request.rb [hunk 5]
```diff
@@ -47,7 +47,11 @@ def build_headers
     end
 
     def transform_header(header)
-      -header[5..].split('_').each(&:capitalize!).join('-')
+      if Gem::Version.new(Rack.release) < Gem::Version.new('3')
+        -header[5..].split('_').each(&:capitalize!).join('-')
+      else
+        -header[5..].tr('_', '-').downcase!
```
Comment (dblock) [inline lib/grape/request.rb]: Don't modify in-place, remove `!`

File: lib/grape/request.rb [hunk 6]
```diff
@@ -47,7 +47,11 @@ def build_headers
     end
 
     def transform_header(header)
-      -header[5..].split('_').each(&:capitalize!).join('-')
+      if Gem::Version.new(Rack.release) < Gem::Version.new('3')
+        -header[5..].split('_').each(&:capitalize!).join('-')
```
Comment (dblock) [inline lib/grape/request.rb]: This was a mistake, should not be modifying in-place either, `capitalize`.
